### ISSUE TITLE
Magical Wood and Reactorcraft

### ISSUE THREAD

**Issue by MissPotato** [rwtema/extrautilities_old_issues#525]
When both Reika's Reactorcraft and Extrautitiles are installed together, the recipe for Magical Wood only yields one no matter the combination of books being used. Even though when I remove Reactorcraft it functions fine. Versions of mods: Reactorcraft V4c, Dragon API V4c, Extrautilities v1.2.2. There is no information in the log that will be useful, but if you want it, I can upload it.

**Comment by ReikaKalseki** [comment]
And the part that confuses me is that I touch neither enchanting, crafting, nor ExU.

**Comment by MissPotato** [comment]
The recipes involving a distinction between stable ingots and unstable ingots fails to make the distinction.  I noticed this when trying to a Builder's Wand using stable ingots (which makes the 
Super Builder's wand.) As well, the stable sigil doesn't make stable ingots. It makes unstable ingots. These all work find without Rotarycraft, some odd stuff is going down.

**Comment by TomeWyrm** [comment]
RotaryCraft or ReactorCraft? Because the magical wood issue is apparently removed by uninstalling ONLY ReactorCraft. Leaving RotaryCraft and DragonAPI installed.

Honestly I should do some testing myself to see if any other combinations cause these issues.

**Comment by MissPotato** [comment]
I'll test it right now and see what happens. It seems like it has something to do with NBT reading as it's happening to other mods than just RWTema's. I'll edit with more information. 

Edit 1: I now have players complaining that they can't even use the unstable ingots on servers but can offline. Though from my understanding the server they're playing on is Cauldron. I'm going to try and reproduce on a vanilla forge server. But after removing mod by mod it stems from Rotary Craft. I don't know if other mods that are dependent on Rotary Craft cause the same issues though because the game won't even start. 

Edit 2: Dorm wifi is pretty awful. Takes forever to download anything. But after setting everything up, their problem is relating to cauldron. 

From what I've gathered thus far: when you craft using Division Sigil it checks to see if an NBT tag exists telling the state of the sigil. 0 being inactive. 1 being unstable. And 2 being Stabilized. The same method exists for ingots. However with ingots only 0 and 1 exists. With Rotary Craft and Dragon API both installed this check isn't made. The same thing happens with bookcases. The crafting recipe checks to the see the level of the enchant and then uses math to calculate the output. However the recipe is being told the enchant levels are "0" and yields a single magical wood.

**Comment by ReikaKalseki** [comment]
ReactorCraft does not touch item NBT either (aside from using it on a couple items), but I am interested in what you might find.

**Comment by MissPotato** [comment]
Thats all the information I've collected thus far. If you have anything more specific to test I'll do so. If there is any tools that will give more information I'll gladly set them up. I've got plenty of spare time after classes.

**Comment by McJty** [comment]
This does sound very similar to what RFTools is experiencing. I also have a crafting recipe where I need to check for an NBT value being exactly the same but also in my case the test is not done and it seems to revert to normal crafting instead (i.e. not my custom recipe).

**Comment by TomeWyrm** [comment]
I just tried the super-builders wand recipe with

CodeChickenCore-1.7.10-1.0.4.35-universal
CoFHCore-[1.7.10]3.0.0RC7-211
DragonAPI 1.7.10 V4c
extrautilities-1.2.2
NotEnoughItems-1.7.10-1.0.4.83-universal
rftools-2.63

Didn't work.

Edit: Magical Wood DOES WORK with that mod list.
As does the Saturation Module from RFTools
Also apparently the Unstable Ingot Block... That _might_ have been because creative mode though.

**Comment by McJty** [comment]
So that's not enough mods to break all the recipes although it does brake the super-builders wand?
Perhaps try adding RotaryCraft and do the same test?

**Comment by MissPotato** [comment]
Yeah the point of my early post was to point out that it's not Dragon API but is specifically Rotary Craft (and possibly extensions of it)

**Comment by TomeWyrm** [comment]
The trick is, DragonAPI alone DOES break the Creative/Super Builder's Wand.

Incidentally RotaryCraft v4c installed does redult in breakage of both the Magical Wood _and_ Saturation Module recipes.

So it likely wasn't ReactorCraft's fault. RC is a RoC addon, and RoC alone breaks both recipes (at least for me). I'm going to try just ExU, RoC, and DragonAPI here in a minute. Removing even NEI (_cry_)

**Comment by TomeWyrm** [comment]
Yes, Only ExU v 1.2.2, RotaryCraft v4c, and DragonAPI v4c results in Magical Wood and Super Builder's Wand breakage.

Ok, it turns out that I'm an idiot and was using the wrong stable ingots with DragonAPI+ExU. It actually DOES work, so now I'm trying with RotaryCraft again. So... yeah.

The Pseudo-Inversion Sigil crafted "stable" Unstable Ingot cannot be crafted with RotaryCraft installed. Which also means the Super Builder's Wand would be unobtainable... Excuse me as I downgrade really quickly to test with an ingot I pull forward. So technically the Super Builder's Wand crafting recipe DOES work with RotaryCraft... you just can't craft the proper kind of unstable ingot.

I have no idea how this works, I blame a hiccup somewhere in the crafting handler(s).

In Summary. All the recipes work with only DragonAPI and the other mods. (ExU, and RFTools+COFHCore). But with RotaryCraft added, the Pseudo-Inversion Sigil crafted stable Unstable Ingot recipe breaks, as does the Magical Wood additional product, and the Saturation Module from RFTools.

**Comment by MissPotato** [comment]
Once I get back from eating I'll see how far back I can roll Rotary Craft (if I can even find older builds) before this issue starts happening.

**Comment by TomeWyrm** [comment]
May I suggest looking for outdated modpacks? A good place to find them is Technic. Luckily Reika has a very consistent naming scheme, and a very nice changelog hosted on his site. I just had to do that for an issue with text in the Lexica Botania.

**Comment by ReikaKalseki** [comment]
> I have no idea how this works, I blame a hiccup somewhere in the crafting handler(s).

I _have_ no crafting handlers...

**Comment by MissPotato** [comment]
@TomeWyrm: and all this time I just thought it was optifine hacking forced text weird. 

@ReikaKalseki Could it be related to the 64+ output method? Not many mods do that, and Rotary Craft is one of them. I doubt it's related but weird stuff happens in minecraft. 

Also, It takes me about 10 minutes to download a 181 mb file through Mediafire, my testing could take a long time.

**Comment by ReikaKalseki** [comment]
> Could it be related to the 64+ output method? Not many mods do that, and Rotary Craft is one of them.

???

**Comment by TomeWyrm** [comment]
Doesn't having a custom crafting block require a custom handler?

In any case it only seems to affect ingredients with NBT data causing special crafting products. Does DragonAPI or RotaryCraft affect, look at, or in any way interact with crafting or NBT handling? I would assume that RotaryCraft will... speaking of, I should flip that config setting and try again. Excuse me.

**Comment by ReikaKalseki** [comment]
> Doesn't having a custom crafting block require a custom handler?

No.

**Comment by MissPotato** [comment]
I'm just kinda spit balling while i test to be honest. But what I'm referencing is when a user puts a 16x Gear Unit into a crafting grid, it returns 244 HSLA Steel Scraps.

**Comment by TomeWyrm** [comment]
I tried true and false on "Crafting Table can Make Machines". No change. Drat.

Actually you NOT having a custom handler might be the issue. Although how that is screwing up NBT detection baffles me. It should be simply checking for an NBT flag to determine the valid item, shouldn't it?

@McJty You actually have an NBT dependent crafting recipe. Do you have any ideas how that could break? Or how we could check to see... wait a second. Let me go grab NBTExplorer and do a bit of tinkering.

**Comment by TomeWyrm** [comment]
Nope. No effect on the items themselves. I need better debugging tools darnit! I have to go do real life chores. Hope you all can get some kind of ideas, because this frankly mystifies me and I can't figure out what broke.
It's not removing the recipes, just incorrectly handling the NBT data during crafting so it outputs the wrong items.

**Comment by MissPotato** [comment]
I can only find version v3d without plugging random modpacks. Though it still exists in v3d if that helps...

**Comment by ReikaKalseki** [comment]
> I'm just kinda spit balling while i test to be honest. But what I'm referencing is when a user puts a 16x Gear Unit into a crafting grid, it returns 244 HSLA Steel Scraps.

Use Minetweaker to remove the recipes for scrap and see what happens.

**Comment by MissPotato** [comment]
Wil do, also stangly, shift clicking the ingots out of the crafting table output make it so they are completely useless. Where mouse1-ing them out allows them to tick.

**Comment by TomeWyrm** [comment]
Try it with v25z for 1.6 and the last relevant version of Extra Utilities for 1.6.4? Because I think I can find a pack or backup with v1 and v2 of RotaryCraft, but if the issue exists back in 1.6 packs I'll work from there… when I get back home

**Comment by MissPotato** [comment]
I remove all recipes that returned scraps and that did not fix the problem. Then I removed rotary craft and dragon api from the mix, and added a recipe that turned one diamond into 128 diamonds. And aside from inventory durps, everything functioned as it is supposed. So it is back to square one I guess.

**Comment by TomeWyrm** [comment]
I'll give it a try with v5 too now that it's out.

**Comment by TomeWyrm** [comment]
Yep, v5 has the same 3 crafting problems as in v4. The Lexica Botania font issue was fixed though, so yay!

Do either @McJty or @ReikaKalseki have any ideas on how to introduce a... crafting-trace to a log file or dump it so we can investigate further?

**Comment by TomeWyrm** [comment]
I can also confirm that this is limited to 1.7 in some way. I tried with the last 1.6.4 versions of NEI, RotaryCraft, and Extra Utilities (and all dependencies also at last 1.6.4 release), the pseudo-inversion ingot and magical wood extra output both work. Obviously RFTools wasn't released for 1.6.4, so I can't test that.

**Comment by McJty** [comment]
In reply to an earlier question. The only way that my recipe can break as far as I see it is if my custom crafting recipe is actually not being used at all. My custom recipe is the NBTMatchingRecipe class. It extends from ShapedRecipes. When the bug is present it appears as if my recipe is just doing what ShapedRecipes does and ignores the additions I made to the subclass. NBTMatchingRecipe extends the methods getCraftingResult() and matches() from ShapedRecipes. The only way I can see this happening is if something changes (with ASM or so) these two methods for all loaded recipes so that my overridden versions are no longer used.

**Comment by ReikaKalseki** [comment]
> In reply to an earlier question. The only way that my recipe can break as far as I see it is if my custom crafting recipe is actually not being used at all. My custom recipe is the NBTMatchingRecipe class. It extends from ShapedRecipes. When the bug is present it appears as if my recipe is just doing what ShapedRecipes does and ignores the additions I made to the subclass. NBTMatchingRecipe extends the methods getCraftingResult() and matches() from ShapedRecipes. The only way I can see this happening is if something changes (with ASM or so) these two methods for all loaded recipes so that my overridden versions are no longer used.

I know Forge goes through and replaces all recipe objects in the main list to use the Ore Dictionary; I have a problem with my own recipe (the ChromatiCraft fragment copying recipe).

**Comment by McJty** [comment]
Hmm, but why does it seem to only do that if RotaryCraft is present? My custom crafter works fine normally. This is really strange...

**Comment by ReikaKalseki** [comment]
Also, all my ASM is available here:
https://github.com/ReikaKalseki/LegacyCraft/blob/master/LegacyASMHandler.java#L88
https://github.com/ReikaKalseki/DragonAPI/blob/master/ASM/DragonAPIClassTransfomer.java#L53
https://github.com/ReikaKalseki/ChromatiCraft/blob/master/Auxiliary/ChromaASMHandler.java#L69

**Comment by TomeWyrm** [comment]
I did just check ChromatiCraft. It doesn't trigger the bug. The only mod I can get to trigger it as of right now is RotaryCraft... 

Let's try this from another angle!
Did you have troubles with any of your crafting recipes in the 1.6.4 to 1.7.10 transition that required you to do something other than just register your recipes like everyone else (I assume) does? How difficult _is_ it to write a custom crafting handler because that could be a way to bypass the bug.

Also can any of you three authors boot RoC+ExU or RoC+RFTools in a dev environment and replicate the bug?

**Comment by McJty** [comment]
In my custom modpack I have both RFTools and RotaryCraft but I'm making dinner right now so can't test right away. I will in a few hours from now.

**Comment by McJty** [comment]
Ok, just tried it and I can confirm that in my custom modpack I also cannot craft the RFTools saturation module.

**Comment by MissPotato** [comment]
I'm not currently at my computer and won't be for some time, I know that Minetweaker has a "WithTag" variable. Could someone see if that breaks too?

**Comment by ReikaKalseki** [comment]
> Did you have troubles with any of your crafting recipes in the 1.6.4 to 1.7.10 transition that required you to do something other than just register your recipes like everyone else (I assume) does? How difficult is it to write a custom crafting handler because that could be a way to bypass the bug.

I do not remember any such cases, and writing a custom crafting handler is a pain in the ass. I tried once, long ago, and promptly gave up and deleted it.

**Comment by ReikaKalseki** [comment]
Still nothing?

**Comment by TomeWyrm** [comment]
Hmm...
Reika, could you try using CodeChickenCore Dev and load up ExtraUtilities or RFTools in your build environment to see if it still breaks there?

If it does still break in a dev environment (because all signs point to something in your code somewhere), you're the author most likely able to safely disable chunks of your code to try and track the bug down.

**Comment by ReikaKalseki** [comment]
ExU refuses to launch in the dev environment, throwing a NoSuchFieldError (SRG field name).

I cannot hunt through my own code without some hint as to the problem. I have several thousand or more lines that may be responsible.

**Comment by TomeWyrm** [comment]
Well the easiest test would be disable all your crafting, wouldn't it? Keep the items ingame, just de-register them?

**Comment by ReikaKalseki** [comment]
Do you mean adding the vanilla-type recipes, or custom crafting tables like the Worktable? I do not see how either could affect it, but this at least is not too difficult to try.

What exactly do I use to determine if it is broken or not? I am completely unfamiliar with magical wood.

**Comment by TomeWyrm** [comment]
Try a high level enchant book. Power V, Sharpness V, etc. The recipe should grant more than one wood per craft if the bug is not present.

Alternately, if RFTools runs in a deobfuscated environment where Extra Utilities does not (You were using CodeChickenCore Dev right? It automatically deobfuscates compiled mods for a dev environment. Way handy) you want to try and craft the Saturation Module. The recipe would be, left-to-right, top-to-bottom: blank, !00% full Zombie Syringe, blank, redstone, gold, redstone, blank, ink sac, blank. The only tricky part is the Syringe. Left-click a zombie with it multiple times until it fills up completely. If the bug is NOT present it will craft the Saturation Module.

**Comment by ReikaKalseki** [comment]
Removing my recipes had no effect, though I have successfully reproduced the bug. I tried opening the ExU source to reverse-engineer his system, but most of his classes display as empty, so that is not going to work, either.

This is not going to get anywhere without input from @rwtema.

**Comment by TomeWyrm** [comment]
Can you reproduce with RFTools? Because McJty is paying attention to this issue.

**Comment by McJty** [comment]
I am? :-)

Let me know if you need any assistance reproducing this

**Comment by ReikaKalseki** [comment]
Can you show me all of the code relating to your crafting handler?

**Comment by McJty** [comment]
All crafting related stuff is here:

https://github.com/McJty/RFTools/tree/master/src/main/java/mcjty/rftools/crafting

The specific handler which is failing to work in this particular case is:

https://github.com/McJty/RFTools/blob/master/src/main/java/mcjty/rftools/crafting/NBTMatchingRecipe.java

**Comment by ReikaKalseki** [comment]
My biggest suspicion is that something is replacing the crafting recipe objects. Have you tried putting console printouts in to ensure your code is getting called?

**Comment by McJty** [comment]
Yes, I have done that test and indeed my code is not called

**Comment by ReikaKalseki** [comment]
Next step is to see everything that calls CraftingManager.recipeList.

....Wait....

I just remembered something I added months ago:
https://github.com/ReikaKalseki/RotaryCraft/blob/master/RotaryRecipes.java#L157

Is that config on and is TE installed?

**Comment by McJty** [comment]
TE is installed and I'm using default RotaryCraft configs. I assume you're talking about this:
    B:"Allow Blast Glass to be Used as TE Hardened Glass"=false

**Comment by ReikaKalseki** [comment]
Yes. Try removing TE and see what happens.

**Comment by McJty** [comment]
Ok. I will try this tomorrow (in 8 hours from now) as I have to go now. Thanks for assisting with this.

**Comment by TomeWyrm** [comment]
TE was not installed for the Extra Utilities test. Or my RFTools test, unless CoFH Core triggers the Blast Glass config.

**Comment by ReikaKalseki** [comment]
No, it would not. So that rules that out...

**Comment by ReikaKalseki** [comment]
Interesting update: Gendustry's gene template crafting, also an NBT handler, works fine.

**Comment by TomeWyrm** [comment]
Does BDew have a dev version or open source?

**Comment by ReikaKalseki** [comment]
I believe Gendustry's source is available.

**Comment by ReikaKalseki** [comment]
Another related (NBT crafting) issue has been reported.

**Comment by McJty** [comment]
What was it? Does it help reproducing this?

**Comment by ReikaKalseki** [comment]
Not really, but it does further confirm what we already knew, that something is replacing NBT-sensitive recipes.

**Comment by MissPotato** [comment]
Is there any testing we can do on a user/noncoder end to give more information?

**Comment by ReikaKalseki** [comment]
Not that I can think of, no.

**Comment by MissPotato** [comment]
Well, I'll keep an eye on the thread, good luck.

**Comment by ReikaKalseki** [comment]
When was the last time @rwtema was seen?

**Comment by TherminatorX** [comment]
@ReikaKalseki On Twitter? 2h ago. On Github? 4 days ago. On this repository actually answering issues? Sometime last year. :/

**Comment by ReikaKalseki** [comment]
:/

**Comment by TomeWyrm** [comment]
This has been fixed!

Apparently it had to do with Forge enforcing the ore dictionary triggering an odd interaction with "a fairly late OreDict.initVanillaEntries which triggers the recipe replacement".

V7b of DragonAPI fixes the issue with Reika's mods, and this issue can provide an example for anyone else somehow triggering the bug. Check to see if Forge is having fun with the Ore Dictionary!

**Comment by Keridos** [comment]
Still bugs my Unstable ingots with the stable sigil, this time it also need TConstruct to be installed.

**Comment by ReikaKalseki** [comment]
How is it different from [this](https://github.com/SlimeKnights/TinkersConstruct/issues/1777), then?
